### ISSUE TITLE
**Primitive definitions**

### ISSUE THREAD

**Issue by ronflima** [cookcup/cookcup#3]
Define all primitive interfaces in order to start building the library.

**Comment by ronflima** [comment]
We have the following to start building this library, in this order:

1. We need to define scopes. Each functionality have a scope to be fired from. 
1. Define functionalities, based on each scope.
1. Define interfaces for each functionality.
1. Define the tests for each interface in order to maintain project's sanity.

I suggest using algorithm argumentation for testing everything due to its effectiveness.

**Comment by ronflima** [comment]
As an annotation, we can rely on this:
```javascript
function c(selector) {
    if (this.constructor == c) {
        alert('called as a class')
        // add here the node creation
    } else {
        alert('called as a function')
        // add here node searching
    }
}
```

This will create a function that can be called as both constructor or function for the _c_ class.

**Comment by ronflima** [comment]
As we have discussed yesterday, we have found this would be a nice syntax for this library:

```javascript
classes(element).add("class1 class2 class3")
classes(element).rm("class1 class2")
classes(element).toggle("class1")
classes(element).has("class1")
```

There are some needed definitions we need to address:

1. _element_ is a selector or a NodeList/NodeElement?
1. Could we chain calls like in:
```javascript
classes(element).toogle("class1").attribute.add("href", "someurl")
```
This fluency is something like some C++ operations but it seems to be a strange way to do things. For instance, methods that return booleans like `classes(element).has("class1")` cannot be chained due to its return type.  I feel that fluency must be achieved through operations within the same class of operations, like with classes. However, even so, there will be limitations for query-like methods like `has`.

**Comment by ronflima** [comment]
Just occurred to me a good insight that could help this library´s design.

Functionalities can be grouped as follows, in my opinion:

- Manipulators, which change DOM elements/structure
- Inquisitors, which query information from DOM elements/structure
- Searchers, which will search for given elements attached to DOM structure

A _manipulator_ will manipulate an existent element, moving it around the DOM or just adjusting/creating attributes. A manipulator can, also, create new elements and attach them to the DOM, if needed.

An _inquisitor_ will query information for a given DOM element. So, they can return information about a given node or list of nodes. 

A searcher will always return an element, collection of elements or nothing, based on a given search criteria. It means that a searcher is used for finding elements within the DOM.

**Comment by caiotarifa** [comment]
@ronflima, replying your points:

> 1. _element_ is a selector or a NodeList/NodeElement?

I understand that a library like this one is willing to make life easier for who respects the JavaScript language and manipulates the DOM. Accepting the three possibilities doesn't break this rule, nor hurt the interface, much less impact performance or size.

So I'd be permissive at this point in order to make the developer's day-to-day life easier.

```javascript
classes('a').add('link')

const element = document.querySelector('a')
classes(element).add('link')

const elements = document.querySelectorAll('a')
classes(elements).add('link')
```

We can also accept common _arrays_, using only items that is a valid element or selector.

```javascript
classes(['a', element, ...elements]).add('link')
```

> 2. Could we chain calls like in:
> ```javascript
> classes(element).toggle("class1").attribute.add("href", "someurl")
> ```

Initially I see no sense for this to work, it hurts the interface and the purpose of this library. Perhaps in the future this may be subject to reassessment, but I see no utility or motives to move on this way.

I'd spend more rows with this problem.

```
const links = document.querySelectorAll('a')

classes(links).toggle('class1')
attributes(links).add('href', 'https://www.example.com')
```

> 3. Manipulators, inquisitors and searcher.

I believe that this division is very clear and we can use it internally. For users, we maintain the simplicity and semantics in the names, categorizing and dividing into initial five modules: **Attributes**, **Classes**, **Events**, **Find** and **Styles**.

**Comment by ronflima** [comment]
# Namespacing

I believe that namespaces are useful for meaningful code as well as for the sake of code organization. It also minimizes names collisions. So we should prefix everything with something like _ccup_. 

Therefore, a call to the library would look like:

```ccup.classes(element).has(“class1”)```

# Other Comments

In my last message I came up with some concepts which are important to build the library, not to use it. So, those are pretty much internal stuff, alright.

**Comment by caiotarifa** [comment]
You're right, `cookcup`, `ccup` or `cc` are valide names to scope (I prefer the smallest), as`$` for jQuery and `_` for Lodash.

In other hand, _ES Modules_ has a great adoption by the front-end developers. So I think it's nice to write a guide for this mode.

```
import cc from 'cookcup'
import { attributes, classes } from 'cookcup'
import classes from 'cookcup/classes'
```
